### PR TITLE
feat: introducing Matcher

### PR DESCRIPTION
**Pull request by sshaman1101**

### CODE REVIEW

File: insonmnia/matcher/matcher.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+}
+
+func (m *matcher) reorderOrders(one, two *sonm.Order) (bid, ask *sonm.Order, err error) {
+	// just a sanity check, orders must have different types
+	if one.GetOrderType() == two.GetOrderType() {
+		return nil, nil, errors.New("orders must have different types")
+	}
+
+	if one.GetOrderType() == sonm.OrderType_BID {
+		bid = one
```
Comment (oopcode) [inline insonmnia/matcher/matcher.go]: can be done in a single line, but whatevs

File: insonmnia/matcher/matcher.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+		bid = two
+	}
+
+	return
+}
+
+func (m *matcher) sortByProfit(orderType sonm.OrderType, orders []*sonm.Order) []*sonm.Order {
+	switch orderType {
+	case sonm.OrderType_BID:
+		// we want to buy most cheapest ASK
```
Comment (oopcode) [inline insonmnia/matcher/matcher.go]: s/most cheapest/cheapest/g

File: insonmnia/matcher/matcher.go [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+	if one.GetOrderType() == sonm.OrderType_BID {
+		bid, ask = one, two
+	} else {
+		ask, bid = one, two
+	}
+
+	return
+}
+
+func (m *matcher) sortByProfit(orderType sonm.OrderType, orders []*sonm.Order) []*sonm.Order {
```
Comment (oopcode) [inline insonmnia/matcher/matcher.go]: this is done by DWH (sorting by price with respect to order type)

File: insonmnia/matcher/matcher.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+	dwhReply, err := m.config.DWH.GetMatchingOrders(ctx, &sonm.MatchingOrdersRequest{
+		Id:    &sonm.ID{Id: id.String()}, // TODO: this is droch. #1 :(
+		Limit: m.config.QueryLimit,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	var orders []*sonm.Order
```
Comment (3Hren) [inline insonmnia/matcher/matcher.go]: You can preallocate this.

File: insonmnia/matcher/matcher.go [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+	// TODO: this is droch #3;
+	bidID, err := util.ParseBigInt(bid.GetId())
+	if err != nil {
+		return nil, err
+	}
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+
```
Comment (3Hren) [inline insonmnia/matcher/matcher.go]: Unneeded blank.

**Pull request merged**